File: Project 2 - How Does it Rank? ND/deck.py
from card import *
from random import shuffle

"""
Module that contains a deck class

:author: Neil Daterao
"""

class Deck: 
    """
    Class that creates a deck object
    """
    
    def __init__(self):
        """
        Constructor that initializes a deck object
        """
        
        self.__deck_contents = []
        for rank in RANKS:
            for suit in SUITS:
                card = Card(rank,suit)
                self.__deck_contents.append(card)
    
    def shuffle(self):
        """
        Function that shuffles the contents of the deck
        
        :return: None
        """
        shuffle(self.__deck_contents)
    
    def deal(self):
        """
        Function that deals a card of the top off the deck, will return None if deck is empty
        
        :return: A card object
        """
        
        if self.size == 0:
            return None
        
        else:
            card_to_be_dealt = self.__deck_contents[self.size()-1]
            self.__deck_contents.pop()
            return card_to_be_dealt
    
    def size(self):
        """
        Function to return the current length of the deck 
        
        :return: Integer representing the length of the deck 
        """
        return len(self.__deck_contents)
    
    
    def __str__(self):
        """
        Return string version of deck with one card per line
        
        :return: String version of the deck with one card per line
        """
        
        string_version_of_deck = ""
        for card in self.__deck_contents:
            string_version_of_deck += str(card) + "\n"
        return string_version_of_deck

if __name__ == "__main__":
    deck = Deck()
    print(deck)
    deck.shuffle()
    print("\n \n \n")
    print(deck)
    print(deck.size())
    print(deck.deal())
    print(deck.size())
    print("\n" ,deck)

File: Project 4 - Java Poker II - The Stud Version/src/proj4/Deck.java
package proj4;

import java.util.ArrayList;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Class that models a standard deck of playing cards
 * 
 * @author Neil Daterao
 */
public class Deck {
    
    private ArrayList<Card> deckOfCards; 
    private int nextToDeal; 
    public static final int MAXCARDAMOUNT = 52; 

    /**
     * Default Constructor. Initializes a standard playing card deck of 52 cards. 
     */
    public Deck() {
        deckOfCards = new ArrayList<Card>(); 
        nextToDeal = 0; 
        for (String suit : Card.SUITS) { 
            int suitAsInt = Card.SUITS.indexOf(suit); 
            
            for (int rank = Card.MINCARDRANK; rank <= Card.MAXCARDRANK; rank++) { 
                deckOfCards.add(new Card(rank, suitAsInt)); 
            }
        }
    }

      /**
     * Method to shuffle the current deck. 
     */
    public void shuffle() { 
        if (nextToDeal < MAXCARDAMOUNT) { 
            for (int currentCardIndex = nextToDeal; currentCardIndex < deckOfCards.size(); currentCardIndex++)  {
                int randomCardIndex = ThreadLocalRandom.current().nextInt(nextToDeal, deckOfCards.size()); 
                Card currentCard = deckOfCards.get(currentCardIndex);
                Card randomCardToSwap = deckOfCards.get(randomCardIndex); 
                deckOfCards.set(currentCardIndex, randomCardToSwap);
                deckOfCards.set(randomCardIndex, currentCard);
    
            }
        }
        else { System.out.println("No cards left in the deck to shuffle!");}
    }

    /**
     * 
     * @return Returns the next undealt card. If all cards have been dealt, will return null.
     * @apiNote This runs in O(1) time since ArrayLists have constant look up times and we're not shifting the index of any cards in the ArrayList. 
     */
    public Card deal(){ 
       if (!isEmpty()) {
            Card cardToBeDealt = deckOfCards.get(nextToDeal); 
            nextToDeal++; 
            return cardToBeDealt; 
       }

       else { return null; }
      
    }

    /**
     * 
     * @return Returns true if deck is empty and false if there are still undealt cards.
     */
    public boolean isEmpty() { 
        return size() == 0;
    }
   
    /**
     * 
     * @return An integer representing the amount of undealt cards in the deck
     */
    public int size() { 
        return MAXCARDAMOUNT - nextToDeal; 
    }
    
    /**
     * 
     * Resets the deck to a state where all cards are undealt. 
     */
    public void gather() {
        nextToDeal = 0; 
    }

    
    /**
     * @return Returns the string version of all the undealt cards in the deck with one playing card per line
     */
    public String toString() { 
        if (nextToDeal < MAXCARDAMOUNT) { 
            String deckOfCardsAsString = ""; 
            
            for (int indexOfCard = nextToDeal; indexOfCard < deckOfCards.size(); indexOfCard++) { 
                deckOfCardsAsString += deckOfCards.get(indexOfCard).toString() + "\n"; 
            }
            
            return deckOfCardsAsString; 
        }
        else { return "Deck is empty!"; }
    }




}
